### PR TITLE
Fetch torch at pip install

### PR DESCRIPTION
**Pull request by MattToast**
This PR is a rivial purposal to unify SmartSim optional dependency resolution to #309. This Pr divides the the optional ML dependencies into 2 categories (`[ml-cpu]` and `[ml-cuida]`) to add the additional context needed to know which version of torch to fetch at `pip install` time.

Similar to #309, `smart build` will continue to warn about ML package conflicts if RAI is built in a different context to the installed  ML packages. A new target to the CLI `smart test` has also been added to verify that a user's install of SmartSim is able to run a simple ML experiment.

Since this PR removes the ability to fetch python packages through pip, the `--only_python_packages` flag of `smart build` has been removed.

### CODE REVIEW

File: smartsim/_core/_cli/test.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+            self._finalizer.detach()  # type: ignore[attr-defined]
+
+
+def execute(_args: argparse.Namespace, /) -> int:
+    backends = installed_redisai_backends()
+    try:
+        with _VerificationTempDir() as temp_dir:
+            test_install(
+                location=temp_dir,
+                port=None,  # TODO: allow users to pass as arg `--port`?
```
Comment (ankona) [inline smartsim/_core/_cli/test.py:76]: I'll steal from Matt and ask if you want to create a task for this and remove the TODO from the code

File: smartsim/_core/_install/builder.py [hunk 2]
```diff
@@ -455,10 +470,10 @@ def build_from_git(self, git_url: str, branch: str, device: str = "cpu") -> None
         # get RedisAI dependencies
         dep_cmd = [
```
Comment (ankona) [inline smartsim/_core/_install/builder.py:471]: could we pull the right hand side of this out into a function and avoid some minor duplication?

```
dep_cmd = create_git_build_cmd(self.fetch_pt, self.fetch_tf, self.fetch_onnx, verbose=True)
build_cmd = create_git_build_cmd(self.torch, self.tf, self.onnx)

File: smartsim/_core/_install/builder.py [hunk 3]
```diff
@@ -309,9 +311,9 @@ def __init__(
         self.rai_install_path: t.Optional[Path] = None
 
         # convert to int for RAI build script
-        self.torch = 1 if build_torch else 0
-        self.tf = 1 if build_tf else 0   # pylint: disable=invalid-name
-        self.onnx = 1 if build_onnx else 0
+        self.torch: _TBoolInt = 1 if build_torch else 0
+        self.tf: _TBoolInt = 1 if build_tf else 0  # pylint: disable=invalid-name
+        self.onnx: _TBoolInt = 1 if build_onnx else 0
```
Comment (ankona) [inline smartsim/_core/_install/builder.py:316]: should we mark these attributes as protected now that we have the public `fetch_xxx` properties to avoid unintended access?

File: smartsim/_core/_install/builder.py [hunk 4]
```diff
@@ -504,7 +519,9 @@ def build_from_git(self, git_url: str, branch: str, device: str = "cpu") -> None
         self.run_command(build_cmd, cwd=self.rai_build_path)
 
         self._install_backends(device)
-        if self.torch:
+        if self.torch and self.torch_dir:
```
Comment (ankona) [inline smartsim/_core/_install/builder.py:522]: would:
```
def user_supplied_backend(backend: str):
  if backend == "torch":
    return self.torch and self.torch_dir
  if backend == "...":
    ...
```
avoid needing comments like the one below?

Comment (MattToast) [reply smartsim/_core/_install/builder.py:522]: YES!! I should have known when I was adding that comment that this is probably a good block of code to isolate into its own fn 🤦

File: smartsim/_core/_install/buildenv.py [hunk 5]
```diff
@@ -312,18 +345,39 @@ def as_dict(self, db_name: DbEngine = "REDIS") -> t.Dict[str, t.Any]:
...
+        def _format_custom_linux_torch_deps(
+            torchv: str, torchvisionv: str, arc: str
+        ) -> t.Tuple[str, ...]:
+            # The correct versions and suffixes were scraped from
+            # https://pytorch.org/get-started/previous-versions/
+            supported_py_versions = ("3.8", "3.9", "3.10")
+            return tuple(
+                itertools.chain.from_iterable(
+                    (
+                        "torch"
```
[the remark this answers is not in the document]
Comment (MattToast) [reply smartsim/_core/_install/buildenv.py:364]: I think it does! Plus it might be nicer to keep this isolated on the change that the links change, as opposed to needing to dive into some obscure `Versioner` method

File: smartsim/_core/_cli/test.py (hunk 1 above)
Comment (MattToast) [reply smartsim/_core/_cli/test.py:76]: damn it, the consequences of my actions are catching up with me :(. Consider it made if/when this goes in!

File: smartsim/_core/_install/builder.py (hunk 2 above)
Comment (MattToast) [reply smartsim/_core/_install/builder.py:471]: I like it! It also removes the need for the `_TBoolInt` type alias. Will add!

File: smartsim/_core/_install/builder.py (hunk 3 above)
Comment (MattToast) [reply smartsim/_core/_install/builder.py:316]: Made them private! I don't think anyone is using these attrs outside of this class, but just in case, I added some read-only `build_*` properties in case someone ever needs to introspect whether or not a backend is going to be built, but removes the ability for them to modify it making the class just a bit more static.

File: smartsim/_core/_cli/cli.py [hunk 6]
```diff
@@ -95,16 +96,22 @@ def default_cli() -> SmartCli:
...
+            site_execute,
+        ),
         MenuItemConfig(
             "clobber",
             "Remove all previous dependency installations",
-            clobber_execute
+            clobber_execute,
+        ),
+        MenuItemConfig(
+            "test",
```
Comment (al-rigazzi) [inline smartsim/_core/_cli/cli.py:112]: This is a great idea! Nevertheless, I wouldn't want users to believe this is our official way of running tests. Can we change it's name to something more restrictive, such as `smoke-test`, `dry-run`, `mini-test`, or similar?

File: smartsim/_core/_cli/test.py [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+    def __exit__(
+        self,
+        exc: t.Optional[t.Type[BaseException]],
+        value: t.Optional[BaseException],
+        tb: t.Optional[TracebackType],
+    ) -> None:
+        if not value:  # Yay, no error! Clean up as normal
+            super().__exit__(exc, value, tb)
+        else:  # Uh-oh! Better make sure this is not implicitly cleaned up
+            self._finalizer.detach()  # type: ignore[attr-defined]
```
Comment (al-rigazzi) [inline smartsim/_core/_cli/test.py:75]: Out of curiosity, what does the scope of `_finalizer` become once you detach it? Does it live until the main thread exits?

File: smartsim/_core/_cli/test.py [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+    exp.generate(orc)
+    exp.start(orc)
+    try:
+        (client_addr,) = orc.get_address()
+        yield Client(address=client_addr, cluster=False)
+    finally:
+        exp.stop(orc)
+
+
+def _find_free_port() -> int:
```
Comment (al-rigazzi) [inline smartsim/_core/_cli/test.py:138]: Interesting approach. Maybe in the future we could think of implementing this for the test ports (in the test suite) too?

File: tests/test_cli.py [hunk 9]
```diff
@@ -668,3 +670,73 @@ def mock_operation(*args, **kwargs) -> int:
...
+def _good_build(*args, **kwargs):
+    print("LGTM")
+
+
+def _bad_build(*args, **kwargs):
+    raise Exception
+
+
+@contextmanager
+def _totally_real_temp_dir(*a, **kw):
```
Comment (al-rigazzi) [inline tests/test_cli.py:684]: OK, so, I don't want to be the killjoy here, but as this is code that _might_ be read by users, I'd use more explanatory names for functions and variables, and I'd apply the same standard to mock values. I know, I know, I'm old.

**Review by al-rigazzi — CHANGES_REQUESTED** [review]
Looks great! Only a couple of minor comments/requests, but I like this new approach (pending future discussions with the rest of the team).

File: smartsim/_core/_cli/test.py [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+
+import numpy as np
+from smartredis import Client
+
+from smartsim import Experiment
+from smartsim._core.utils.helpers import installed_redisai_backends
+from smartsim._core._cli.utils import SMART_LOGGER_FORMAT
+from smartsim.log import get_logger
+
+SMART_LOGGER_FORMAT = "[%(name)s] %(levelname)s %(message)s"
```
Comment (MattToast) [inline smartsim/_core/_cli/test.py:43]: This assignment is shadowing the previous import

File: smartsim/_core/_cli/build.py [hunk 11]
```diff
@@ -30,62 +30,35 @@
...
-                                       SMART_LOGGER_FORMAT)
+from smartsim._core._cli.utils import color_bool, SMART_LOGGER_FORMAT
 from smartsim._core._install import builder
 from smartsim._core._install.buildenv import (
     BuildEnv,
     SetupError,
     Version_,
     Versioner,
     DbEngine,
+    VersionConflictError,
```
Comment (ashao) [inline smartsim/_core/_cli/build.py:43]: Just double checking, `VersionConflictError` and `SetupError` are both here because they get used inside of `buildenv` too (which we're trying to keep 100% separated from all SmartSim code)?

File: smartsim/_core/_cli/build.py [hunk 12]
```diff
@@ -208,33 +164,17 @@ def build_redis_ai(
...
-                sys.exit(1)
-        else:
-            # install pytorch wheel, and get the path to the cmake dir
-            # we will use in the RAI build
-            install_torch(build_env, versions, device=device)
-            torch_dir = build_env.torch_cmake_path
+    if use_torch and torch_dir:
+        torch_dir = Path(torch_dir).resolve()
+        if not torch_dir.is_dir():
+            raise SetupError("Could not find requested user Torch installation")
```
Comment (ashao) [inline smartsim/_core/_cli/build.py:171]: Print the `torch_dir` string too

File: smartsim/_core/_cli/build.py [hunk 13]
```diff
@@ -285,124 +225,181 @@ def build_redis_ai(
...
-    backend_torch_path = f"{CONFIG.lib_path}/backends/redisai_torch"
-    device = "cpu"
-    if Path(f"{backend_torch_path}/lib/libtorch_cuda.so").is_file():
-        device = "gpu"
-    return device
+def check_py_torch_version(versions: Versioner, device: _TDeviceStr = "cpu") -> None:
+    """Check Python environment for TensorFlow installation"""
+
+    if BuildEnv.is_macos():
+        device_suffix = ""
```
Comment (ashao) [inline smartsim/_core/_cli/build.py:232]: For paranoia, throw a warning (at least) here if someone requests gpu on a Mac. Or maybe do it in the extras vehcking stage.

File: smartsim/_core/_cli/cli.py [hunk 14]
```diff
@@ -95,16 +96,22 @@ def default_cli() -> SmartCli:
...
+            "Run a simple SmartSim experiment to cofirm that it is built correctly",
```
Comment (ashao) [inline smartsim/_core/_cli/cli.py:113]: 'cofirm' -> 'confirm'

File: smartsim/_core/_cli/test.py [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+        ],
+        name="FullyConnectedNetwork",
+    )
+    fcn.compile(
+        optimizer="adam", loss="sparse_categorical_crossentropy", metrics=["accuracy"]
+    )
+    model, inputs, outputs = serialize_model(fcn)
+
+    client.set_model(
+        "keras-fcn", model, "TF", device="CPU", inputs=inputs, outputs=outputs
```
Comment (ashao) [inline smartsim/_core/_cli/test.py:166]: For all these backend tests, let's add an optional flag to `validate`, `--device` to specify whether we're testing CPU or GPU. GPU is almost always where we run into trouble

**Review by ashao — CHANGES_REQUESTED** [review]
Some small changes, with maybe the request to add `--device` to a `smart validate` being the one to require a small review

File: smartsim/_core/_cli/cli.py (hunk 6 above)
Comment (MattToast) [reply smartsim/_core/_cli/cli.py:112]: I thought there was a proposal to change it to `test` to mimic the gmake pattern of `make install; make test`. More than willing to revert to `smart validate` though!

File: tests/test_cli.py (hunk 9 above)
Comment (MattToast) [reply tests/test_cli.py:684]: Hahaha, tbf this is absolutely a name I will regent the literally the next time I have to edit this test. Consider it changed to something reasonable!

File: smartsim/_core/_cli/test.py (hunk 8 above)
Comment (MattToast) [reply smartsim/_core/_cli/test.py:138]: I'd be more than willing to add something like this to the test suite! It's defiantly not the most robust solution, but certainly should be "good enough" for our test suite!

I'll write up a ticket!!

File: smartsim/_core/_cli/build.py (hunk 11 above)
Comment (MattToast) [reply smartsim/_core/_cli/build.py:43]: Exactly! The `buildenv` module needs to be imported into the `setup.py` which means it any imports in that module would need to be done through the `importlib` module which can be kind of a pain. As much as I don't love that `VersionConflictError` does not live in the `errors` module, I do think this is the lesser of two evils.

File: smartsim/_core/_cli/build.py (hunk 12 above)
Comment (MattToast) [reply smartsim/_core/_cli/build.py:171]: Smart!!

File: smartsim/_core/_cli/build.py (hunk 13 above)
Comment (MattToast) [reply smartsim/_core/_cli/build.py:232]: good call!

File: smartsim/_core/_cli/cli.py (hunk 14 above)
Comment (MattToast) [reply smartsim/_core/_cli/cli.py:113]: Spelling is hard...

File: smartsim/_core/_cli/test.py (hunk 15 above)
Comment (MattToast) [reply smartsim/_core/_cli/test.py:166]: I'm a little worried about feature creep as this isn't directly related to moving the torch deps. Should we put this into this PR? Would you mind if I kicked this suggestion out into the [`--port` flag](https://github.com/CrayLabs/SmartSim/pull/336/files/ac4734676050a5866d9b1a7013e66a1993d8277f#diff-0f39699c06b8555c21a533e5b264ca4ffc290d50284859a3e1710d1a89f1304c) ticket previously surfaced?

File: smartsim/_core/_install/buildenv.py [hunk 16]
```diff
@@ -333,7 +405,26 @@ def ml_extras_required(self) -> t.List[str]:
...
+                *common,
+                # osx
+                f'torch=={self.TORCH} ; sys_platform == "darwin"',
+                f'torchvision=={self.TORCHVISION} ; sys_platform == "darwin"',
+                # linux
+                *_format_custom_linux_torch_deps(
+                    self.TORCH, self.TORCHVISION, self.TORCH_CPU_SUFFIX.lstrip("+")
+                ),
+            ],
+            "ml-cuda": [
```
[the remark this answers is not in the document]
Comment (MattToast) [reply smartsim/_core/_install/buildenv.py:420]: I'm a little nervous about introducing a `pip install smartsim[ml-rocm]` optional deps without also introducing a `smart build --device rocm` target, even if it is possible to do via the `smart build --torch_dir=/path/to/rocm/torch/...`.

That said I don't have a super strong opinion here and could be convinced to add it. Thoughts from other reviewers (@ashao)?

File: smartsim/_core/_install/builder.py [hunk 17]
```diff
@@ -453,17 +480,20 @@ def build_from_git(self, git_url: str, branch: str, device: str = "cpu") -> None
...
-            f"WITH_ORT={self.onnx}",
-            "VERBOSE=1",
-            self.binary_path("bash"),
-            str(self.rai_build_path / "get_deps.sh"),
-            str(device),
-        ]
+        dep_cmd = self._rai_build_env_prefix(
+            with_pt=self.build_torch,
+            with_tf=self.build_tf,
+            with_ort=self.build_onnx,
```
Comment (ankona) [inline smartsim/_core/_install/builder.py:486]: very nice pivot from the hardcoded strings!

File: smartsim/_core/_cli/test.py (hunk 1 above)
Comment (MattToast) [reply smartsim/_core/_cli/test.py]: Other additional arguments were requested for the `validate` target. A `--port` was additionally added in d9316f595f099a8ee79c0dd042bdca8acaa7532d

File: smartsim/_core/_cli/test.py (hunk 15 above)
Comment (MattToast) [reply smartsim/_core/_cli/test.py]: NVM, I ended up throwing something together in d9316f595f099a8ee79c0dd042bdca8acaa7532d and e1720e1d5ceb876ba2ccc2a602e8d20220122d39. You should now be able to validate against a GPU

File: smartsim/_core/_install/builder.py (hunk 17 above)
Comment (MattToast) [reply smartsim/_core/_install/builder.py:486]: I'd love to say I came up with the idea, but honestly I was just following the advice of a very wise reviewer, hahaha!!

File: smartsim/_core/_cli/test.py (hunk 7 above)
Comment (al-rigazzi) [reply smartsim/_core/_cli/test.py]: Lovely, thanks for the explanation.

File: smartsim/_core/_install/buildenv.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply smartsim/_core/_install/buildenv.py:420]: OK, so, just triple-checking here: a user who wanted to use AMD GPUs would still be capable of doing so by passing `--torch_dir=/path/to/amd_version_of_torch`, correct?

[the remark this answers is not in the document]
Comment (MattToast) [reply smartsim/_core/_install/buildenv.py:420]: That is correct!

[the remark this answers is not in the document]
Comment (MattToast) [reply smartsim/_core/_install/buildenv.py:420]: Similarly users could also still independently install AMD GPU compatible versions of the torch python package deps via `pip3 install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/rocmX.Y.Z`.

I just don't want to give ROCm the illusion of first class support if it is not yet supported to the extent that CPU/CUDA is currently

**Review by al-rigazzi — APPROVED** [review]
LGTM, thanks for taking the time of adding all those options!

File: smartsim/_core/_install/buildenv.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply smartsim/_core/_install/buildenv.py:420]: Grazie
